src/App.jsx

```diff
@@ -20,31 +109,124 @@ function App() {
         [...prevState, { key: data.key, val: data.val() }]
       );
     });
+
+    // shows - edit message
+    onChildChanged(messagesRef, (data) => {
+      // Update the message in local state
+      setMessages((prevMessages) =>
+        prevMessages.map((message) =>
+          message.key === data.key
+            ? { key: data.key, val: data.val() }
+            : message
+        )
+      );
+    });
+
+    // show that the message has been removed
+    onChildRemoved(messagesRef, (data) => {
+      // Remove the message from local state
+      setMessages((prevMessages) =>
+        prevMessages.filter((message) => message.key !== data.key)
+      );
+    });
   }, []);
 
-  const writeData = () => {
-    const messageListRef = ref(database, DB_MESSAGES_KEY);
-    const newMessageRef = push(messageListRef);
-    set(newMessageRef, "abc");
+  const writeData = (e) => {
+    e.preventDefault();
+
+    //images
+    const fileRef = storageRef(
+      storage,
+      `${IMAGE_FOLDER}/${userInputFile.name}`
+    );
+
+    //upload images
+    uploadBytes(fileRef, userInputFile)
+      .then(() => getDownloadURL(fileRef))
+      .then((downloadURL) => {
+        //messages
+        const messageListRef = ref(database, DB_MESSAGES_KEY);
+        const newMessageRef = push(messageListRef);
+
+        const messageData = {
+          text: userInput,
+          timestamp: DATE,
+          imageURL: downloadURL,
+        };
+
+        set(newMessageRef, messageData);
+        setUserInput("");
+        setUserInputFile("");
+        setFileInputValue("");
+        console.log("Image URL:", downloadURL);
+      })
+      .catch((error) => {
+        // Handle errors during image upload or URL retrieval
+        console.error("Error uploading image or getting URL:", error);
+      });
   };
 
-  // Convert messages in state to message JSX elements to render
-  let messageListItems = messages.map((message) => (
-    <li key={message.key}>{message.val}</li>
-  ));
+  //router
 
```

Comment:
Nice implementation of the RequireAuth component and React Router Dom. Normally I wouldnt make RequireAuth a component, more like a redirect, but this will work just as well due to the way you set up the React Router. 
You could try to have multiple pages, and alter requireAuth to something more similar to this: https://bc.rocketacademy.co/2-full-stack/2.2-advanced-react/2.2.2-react-router#private-routing